bootloader/lk.go

```diff
@@ -0,0 +1,202 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package bootloader
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/bootloader/lkenv"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+)
+
+type lk struct{}
+
+// newLk create a new lk bootloader object
+func newLk() Bootloader {
+	e := &lk{}
+	if !osutil.FileExists(e.envFile()) {
+		return nil
+	}
+
+	return e
+}
+
+func (l *lk) Name() string {
+	return "lk"
+}
+
+func (l *lk) Dir() string {
+	// we have two scenarios, image building and runtime
+	// during image building we store environment into file
+	// at runtime environment is written directly into dedicated partition
+	if dirs.GlobalRootDir == "/" {
+		return "/dev/disk/by-partlabel/"
+	} else {
+		return filepath.Join(dirs.GlobalRootDir, "/boot/lk/")
+	}
+}
+
+func (l *lk) ConfigFile() string {
+	return l.envFile()
+}
+
+func (l *lk) envFile() string {
+	// as for dir, we have two scenarios, image building and runtime
+	if dirs.GlobalRootDir == "/" {
+		// TO-DO: this should be eventually fetched from gadget.yaml
+		return filepath.Join(l.Dir(), "snapbootsel")
```

Comment:
I think something like http://paste.ubuntu.com/p/46FCtnfFX2/ is easier to follow. Basicly this makes ConfigFile always look at the on-disk config file. And the detection of LK is done in newLk without the use of ConfigFile. For most boot loaders it make sense to just use "if FileExists(ConfigFile()))" - however because LK will put this into a partition I think we need to separate detection of LK and config file handling. I hope I do not miss something

Reply:
More like this: http://paste.ubuntu.com/p/qJjCzjYSjW/

Reply:
Btw, I wonder how the /boot/lk/snapbootsel.bin becomes the snapbootsel partition - also - do we need ConfigFile() at image build time at all?

Reply:
I.e. could we simply *not* install /boot/lk/snapbootsel.bin (because its not really used, just the partition is used)?

Reply:
It actually operates in quite same way as other bootloaders here. There are two files 

> Btw, I wonder how the /boot/lk/snapbootsel.bin becomes the snapbootsel partition - also - do we need ConfigFile() at image build time at all?

We do need `snapbootsel.bin` as this is initial boot configuration for device to boot at all. It works same way as for u-boot. We generate initial uboot.env at image build time ( which is based on default env file from gadget with added `snap_kernel` `snap_core` values.
Same way operates `snapbootsel.bin` only difference is that image builder instead of putting file to `boot` partition, it uses file as image for boot select partition.
This is also paving way for `uboot.env` being separate partition. You need to generate correct `uboot.env` file at image build time, which is used by image builder as partition image, and at run time partition is used instead.